Compiler warning fixes

jeffhammond opened the pull request:
Please review these carefully, as I have not performed any testing yet.

I don't know if Mac is even supported, but merely trying to compiler revealed one major issue with `thread_t`, which I placed in the `fgmpi_` namespace.

The other commits are relatively trivial changes to silence compiler warnings.  They are not necessarily the best solutions, but perhaps the easiest.

I will eventually setup PRK testing against the FG-MPI Github repo rather than the release tarball, which will give me slightly more confidence that my contributions are reliable.

jeffhammond committed:
add namespace to thread_t

Mac OSX defines thread_t already...

This is the compiler error:

Making all in .
  CC       src/mpi/attr/lib_libpmpi_la-attr_delete.lo
In file included from ../src/mpi/attr/attr_delete.c:8:
In file included from ../src/include/mpiimpl.h:73:
In file included from ../src/include/mpiutil.h:14:
In file included from ./src/util/timers/mpiu_timer.h:63:
In file included from /usr/include/mach/mach_time.h:32:
/usr/include/mach/mach_types.h:119:22: error: typedef redefinition with
different types ('mach_port_t' (aka 'unsigned int') vs
      'struct thread_st')
typedef mach_port_t             thread_t;
                                ^
/Users/jrhammon/Work/MPI/FG-MPI/git/src/fgutil/src/threadlib.h:16:26:
note: previous definition is here
typedef struct thread_st thread_t;
                         ^
1 error generated.
make[2]: *** [src/mpi/attr/lib_libpmpi_la-attr_delete.lo] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2

jeffhammond committed:
remove inline from definitions where it breaks build

inline functions do not necessarily have symbols, which causes linking
to break.  this error was observed with the following compiler:

$ clang --version
Apple LLVM version 7.0.2 (clang-700.1.81)
Target: x86_64-apple-darwin14.5.0
Thread model: posix

jeffhammond committed:
add header to declare MPIDI_CH3U_Buffer_free

This is the compiler warning:

warning: implicit declaration of function 'MPIDI_CH3U_Buffer_free' is
invalid in C99

Adding mpidimpl.h solves this problem.  Another solution would be to
wrap this function to create something like MPIR_Buffer_free.

jeffhammond committed:
silence two undeclared function warnings

- declare MPIU_Get_intra_osproc_rank where needed in collectives
  routines.

- add MPIR_Progress_wait_send_request to mpiimpl.h next to the other
  MPIR_Progress* function.

jeffhammond committed:
silence compiler warning related to non-const assignment

This was the compiler warning:

warning: initializing 'void *' with an expression of type 'const void *'
discards qualifiers

The solution was to declare "const void * buf" instead of "void * buf".

humairakamal commented:
Thanks for your fixes especially for prefixing fgmpi_ to thread_t. I had tested mainly on Linux systems.
I have added two comments in (https://github.com/jeffhammond/fgmpi/commit/8ed5d71161cacf280f50e50231b53f47d5499580) and (https://github.com/jeffhammond/fgmpi/commit/41ada824a808feed6a520c6f9b9bf971672afd73)

jeffhammond committed:
include mpidimpl.h at end of mpiimpl.h for pt2pt

it must be included at the end in order to capture all the stuff
before.  including it with the other FGMPI headers or even after
all the other headers does not work.

this is probably the _wrong_ way to do this, and we should instead
try to expose MPIDI_CH3U_Buffer_free via some new MPID function such
as MPID_Buffer_free, but that is TODO.

jeffhammond committed:
preprocessor guard on FGMPI change

jeffhammond commented:
I addressed both comments.

There are still plenty of undeclared functions that I need to address...

jeffhammond committed:
proper namespacing of FGMPI utility functions

jeffhammond committed:
remove inline in def/decl of symbol used externally
